core/services/keeper/orm.go

```diff
@@ -51,40 +60,42 @@ func (korm ORM) UpsertUpkeep(ctx context.Context, registration *UpkeepRegistrati
 		WithContext(ctx).
 		Clauses(clause.OnConflict{
 			Columns:   []clause.Column{{Name: "registry_id"}, {Name: "upkeep_id"}},
-			DoUpdates: clause.AssignmentColumns([]string{"execute_gas", "check_data"}),
+			DoUpdates: clause.AssignmentColumns([]string{"execute_gas", "check_data", "positioning_constant"}),
 		}).
 		Create(registration).
 		Error
 }
 
-func (korm ORM) BatchDeleteUpkeeps(ctx context.Context, registryID int32, upkeedIDs []int64) error {
-	return korm.DB.
-		WithContext(ctx).
-		Where("registry_id = ? AND upkeep_id IN (?)", registryID, upkeedIDs).
-		Delete(UpkeepRegistration{}).
-		Error
-}
-
-func (korm ORM) DeleteRegistryByJobID(ctx context.Context, jobID int32) error {
+func (korm ORM) BatchDeleteUpkeepsForJob(ctx context.Context, jobID int32, upkeedIDs []int64) error {
 	return korm.DB.
-		WithContext(ctx).
-		Where("job_id = ?", jobID).
-		Delete(Registry{}). // auto deletes upkeep registrations
-		Error
+		WithContext(ctx).Exec(
+		`DELETE FROM upkeep_registrations WHERE registry_id = (
+			SELECT id from keeper_registries where job_id = ?
+		) AND upkeep_id IN (?)`,
+		jobID,
+		upkeedIDs,
+	).Error
 }
 
-func (korm ORM) EligibleUpkeeps(ctx context.Context, blockNumber int64) (upkeeps []UpkeepRegistration, _ error) {
+func (korm ORM) EligibleUpkeeps(
+	ctx context.Context,
+	blockNumber int64,
+	gracePeriod int64,
+) (upkeeps []UpkeepRegistration, _ error) {
+	lastValidBlockHeight := blockNumber - gracePeriod
```

Comment:
@samsondav Should be fine here! This will only happen if we're on a new (maybe local test) chain or if someone sets the gracePeriod to a massive number. Either way, if your grace period is 1M blocks, and the chain is only on block 25... you have to mine 999,975 before your upkeeps become elligible. You're getting what you asked for.

Reply:
actually, now that I think about it... this probably needs a special case for lastValidBlockHeight < 1. Because we use `last_run_block_height` = 0 to mean "this has never been run", so we want to go ahead and call it regardless of `lastValidBlockHeight`. So I think `lastValidBlockHeight` should be min 1.